### ISSUE TITLE
Poor BigInt support

### ISSUE THREAD

**Issue by kryptus36** [supabase/supabase#22338]
This is a bit of a meta-bug, as it were. This issue seems to be swept under the rug as no big deal, but the crux of the issue here is BigInt, or int8, is a base postgres type that breaks in numerous ways when trying to use it in supabase. 64 bit numbers have beein in common use for ~20 years.

BigInt is used in numerous id schemes as it indexes well, can be used as a sort mechanism, and converted to convenient strings for URLs. It is widely used in Supabase own docs as an id column. But if you try to use it, your app can break in confusing ways (no error is returned, and id is incorrect) or you have to use annoying workarounds (convert your id to a string then to a BigInt on every select? Really?)
https://stackoverflow.com/questions/74549311/how-to-use-bigint-safely-with-supabase-js

The BigInt type is supported by every browser as well as Typescript.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-2.html#bigint

And yet when you generate types with supabase it generates them as number.
https://github.com/supabase/supabase-js/issues/863#issuecomment-1835478711

And if you select one larger than Number.MAX_SAFE_INTEGER it will get rounded. This affects the dashboard as well. Despite bugs getting fixed to allow you to input a BigInt in the UI, and sort by them, it will still DISPLAY them incorrectly. No errors, just inaccurate data.

Despite it, again, being a base type in postgres, it is frustrating to many users as is evidenced by the multitude of bugs in various states related.

https://github.com/supabase/supabase-js/issues/863
https://github.com/supabase/postgrest-js/issues/319
https://github.com/supabase/postgrest-js/issues/273
https://github.com/supabase/postgrest-js/issues/419

And a host of others closed as duplicates. I recognize this will get immediately closed, but I hope it draws attention to a gaping hole in supabase tools.

**Comment by encima** [comment]
Thanks for opening. I'll close this as you mentioned expecting immediate closure 😛 

As I understand it, this is often an issue with the support for large numbers with Typescript/Javascript but it is certainly something we can add guardrails for and improve the DX. 

Thanks for linking the existing issues here also, it would be super useful if you could help to narrow down some key areas to work on and I can see what we can do?

For example, if the dashboard handled X and the SDK did Y then we could consider the issue closed.

**Comment by kryptus36** [comment]
Thanks for reading my essay :+1: 

The main issue is if you use the Supabase client to select a column with an int8 value > Number.MAX_SAFE_INTEGER it will get rounded. These numbers are common in many time based ID schemes such as https://www.foxhound.systems/blog/time-sorted-unique-identifiers/

For example, if my database contains the id 561242882974035048 the supabase client will select it as a javascript number 561242882974035100. The last 3 digits are different because the normal number type in JS doesn't go that high.

Fixes:
-The Supabase client needs to cast numbers stored in PG as int8/bigint  to BigInt internally.
-If the column type in PG is int8/bigint the type generator should type to BigInt, not number
-Apparently Postgrest.js suffers from the same issue.

I honestly tried to go do it myself, but couldn't find the relevant code. If someone points me in the right direction, I'll do my best to create a PR. It's *probably* not that hard for someone familiar with the js clients.

**Comment by Taylor123** [comment]
Running into the same issue. Would really like to see some movement here as this can break many applications.

Supabase/postgREST need to auto convert to BigInt or strings if the number is too large.

**Comment by waynesbrain** [comment]
I like the way [node-pg](https://github.com/brianc/node-pg-types/blob/master/lib/textParsers.js#L67) just returns a `string` and then they let you [set a parser](https://github.com/brianc/node-pg-types?tab=readme-ov-file#use) if you want to convert to number everywhere.

**Comment by akiszka** [comment]
I'd like to offer my perspective as I work at a company that handles a lot (>100GB) of financial data in Supabase. We use `numeric` columns to support storing numbers of arbitrary precision, sometimes >2^128. In addition we use `int8` to store ids, and some other random small numbers.

To me, auto-converting values exceeding `Number.MAX_SAFE_INTEGER` seems like the worst of both worlds. You don't get the simplicity of only working with `Number` types, and you can't avoid doing a million ugly type conversions, since you have to cast everything to `BigInt` anyway when you want to do any kind of math operation.

Not to mention that a lot of people (including us) use Postgres' `int8` for columns that never exceed 2^53-1, and changing the types generated by Supabase would force them to rewrite a lot of their code. Even if the feature was opt-in, we would have to deal with the burden of every id being `bigint` or `bigint | number`.

**My preferred solution would be to extend https://github.com/supabase/postgrest-js to support casting columns to bigint, like this:**

```ts
const { data } = await supabase.from("table").select("column::bigint").single();
assert(typeof data.column === "bigint");
```

This would allow us to fetch large numbers from the database without any additional typechecks or conversions. At the same time, we can keep default behaviours in place, and not break the "text" workaround:

```ts
const { data: dataNormal } = await supabase.from("table").select("column").single();
assert(typeof dataNormal.column === "number");

const { data: dataText } = await supabase.from("table").select("column::text").single();
assert(typeof dataText.column === "string");
```

If people like this change, I can implement this in the postgrest client, since this would make my life a lot easier :)

**Comment by waynesbrain** [comment]
> To me, auto-converting values exceeding `Number.MAX_SAFE_INTEGER` seems like the worst of both worlds.

Agreed @akiszka and I would also prefer to just have a number. Currently, for my own `bigint` surrogate id columns, I set MAXVALUE to `Number.MAX_SAFE_INTEGER` in `GENERATED ALWAYS AS IDENTITY` statements, e.g.

```sql
CREATE TABLE public.things (
  id BIGINT NOT NULL PRIMARY KEY GENERATED ALWAYS AS IDENTITY ( MAXVALUE 9007199254740991 ),
  title TEXT
);
```

If I wanted to store `bigint` data that was generated externally, I would decide what type to store it as based on whether I cared about it's numberness in queries or not. I'd probably store it as `text` if it's a key to an external system. If it's a measurement or something I'd store it as `bigint` and then do conversions where necessary.

**Comment by kryptus36** [comment]
@akiszka I think your proposal sounds reasonable. I stopped using supabase because of this and another issue I couldn't resolve.

To be clear, I wasn't proposing to only convert big numbers to bigint, but anything stored in an int8 column. Because accuracy matters. JS sucks for dealing with big numbers, unfortunately.

I do think that if returning as a number > Number.MAX_SAFE_INTEGER that is not cast it should raise a warning in the JS client. As of my last testing, it was silently rounded and that was VERY confusing.

The solution here needs to avoid ambiguity in either case. People wanting bigint need to be able to tell the type generator & postgrest that's what they want / expect. It also needs to be documented well. I wasted a lot of time on this, as have many others. Definitely left a bad impression for me, unfortunately.

**Comment by don4of4** [comment]
Hey all, there is another part of this issue.  

When defining a function which requires a bigint:
```
- Create function to add message and update ledger
CREATE OR REPLACE FUNCTION public.add_message_and_update_ledger(
  p_sender_id BIGINT,
  p_thread_id BIGINT,
)
RETURNS INT
LANGUAGE plpgsql
SECURITY DEFINER
SET search_path = ''
AS $$
...
```

Supabase will generate a type which isn't compatible:
```
...
  Functions: {
      example_function: {
        Args: {
          p_sender_id: number
          p_thread_id: number
        }
        Returns: number
      }
...
```

So when you typecheck you get an error: `Type 'bigint' is not assignable to type 'number'.ts(2322)`

Fundamentally, I think we should generate types for functions which are using BigInt where appropriate, right?  CC @kryptus36 @akiszka

**Comment by jens-f** [comment]
I ran into a similar issue with functions and posted a proposal in the supabase js repo here: https://github.com/supabase/postgrest-js/issues/319
Linking here, in case this repo is the better place to discuss.

**Comment by devjoseh** [comment]
I'm experiencing a similar problem. When I use supabase cli to generate the types file, whatever is bigint is simply set to number. The only solution I've found is to manually change whatever is number to bigint within the types file, but this is definitely not good practice. Same problem as @don4of4

**Comment by Hallidayo** [comment]
Hi all, We’re going to close this one due to inactivity. Please feel free to reopen if you’re still having issues here or open a new issue.

**Comment by theSherwood** [comment]
This is still a problem. Is there any intent to fix this?

**Comment by jens-f** [comment]
@Hallidayo we also still have that issue, it was never addressed as far as I can tell. I'd appreciate it if somebody could have a look at my proposal at https://github.com/supabase/postgrest-js/issues/319#issuecomment-2564090839

**Comment by RampantDespair** [comment]
I'm not sure I understand, this was closed as inactive but completed?
I'm still getting rounded down values when surpassing the max integer value, ie:
`9007199254740993` becomes `9007199254740992`
